appinventor/components/src/com/google/appinventor/components/runtime/ChatBot.java

```diff
@@ -0,0 +1,488 @@
+      }
+      });
+  }
+
+  private void performRequest(String uuid, String question) {
+    // languageToTransateTo is provided either as a two letter code, or two
+    // two letter codes separated by a dash. If only one two letter code is
+    // provided, it is the target language and we set the source language to auto
+    // which tells the service to detect the language
+    HttpsURLConnection connection = null;
+    ensureSslSockFactory();
+    try {
+      byte [] decodedToken = Base58Util.decode(apiKey);
+      ChatBotToken.token token = ChatBotToken.token.parseFrom(decodedToken);
+      ChatBotToken.request request;
+      if (system.equals("") || !uuid.equals("")) {  // No system string
+        request = ChatBotToken.request.newBuilder()
+          .setToken(token)
+          .setUuid(uuid)
+          .setQuestion(question).build();
+      } else {
+        request = ChatBotToken.request.newBuilder()
+          .setToken(token)
+          .setUuid(uuid)
+          .setSystem(system)
+          .setQuestion(question).build();
+      }
+      URL url = new URL(CHATBOT_SERVICE_URL);
+      connection = (HttpsURLConnection) url.openConnection();
+      if (connection != null) {
+        try {
+          connection.setSSLSocketFactory(sslSockFactory);
+          connection.setRequestMethod("POST");
+          connection.setDoOutput(true);
+          request.writeTo(connection.getOutputStream());
+          final int responseCode = connection.getResponseCode();
+          ChatBotToken.response response = ChatBotToken.response.parseFrom(connection.getInputStream());
+          String returnText;
+          if (responseCode == 200) {
+            returnText = response.getAnswer();
+            this.uuid = response.getUuid();
+          } else {
+            returnText = getResponseContent(connection, false);
+          }
+          final String chatbotorerror = returnText;
+
+          // Dispatch the event.
+          GotResponse(responseCode + "", chatbotorerror);
+        } finally {
+          connection.disconnect();
+        }
+      }
+    } catch (Exception e) {
+      e.printStackTrace();
+      if (e instanceof FileNotFoundException && connection != null) {
+        String returnText;
+        try {
+          returnText = getResponseContent(connection, true);
+        } catch  (IOException ee) {
+          returnText = "Error Fetching Translation";
+        }
+        GotResponse("404", returnText);
+      } else {
+        GotResponse("400", "Error Fetching Translation");
+      }
+    }
+  }
+
+  /**
+   * Event indicating that a request has finished and has returned data (translation).
+   *
+   * @param responseCode the response code from the server
+   * @param translation the response content from the server
+   */
+  @SimpleEvent(description = "Event fired when the Chat Bot answers a question.")
+  public void GotResponse(final String responseCode, final String responseText) {
+    activity.runOnUiThread(new Runnable() {
```

Comment:
If we use `form` here (inherited from AndroidNonvisibleComponent), we can avoid having to store `activity` as a separate field.

Reply:
done.